homeassistant/helpers/script.py

```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
     """Throw if script needs to suspend."""
 
 
+class _ScriptRunBase(ABC):
+    """Common data & methods for managing Script sequence run."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        script: "Script",
+        variables: Optional[Sequence],
+        context: Optional[Context],
+    ) -> None:
+        self._hass = hass
+        self._script = script
+        self._variables = variables
+        self._context = context
+        self._step = -1
+        self._action: Optional[Dict[str, Any]] = None
+
+    def _changed(self):
+        self._script._changed()  # pylint: disable=protected-access
+
+    @property
+    def _config_cache(self):
+        return self._script._config_cache  # pylint: disable=protected-access
+
+    @abstractmethod
+    async def async_run(self) -> None:
+        """Run script."""
+
+    async def _async_step(self):
+        try:
+            await getattr(self, f"_async_{_determine_action(self._action)}_step")()
+        except Exception as err:
+            if not isinstance(err, (_SuspendScript, _StopScript)):
+                self._log_exception(err)
+            raise
+
+    @abstractmethod
+    async def async_stop(self) -> None:
+        """Stop script run."""
+
+    def _log_exception(self, exception):
+        action_type = _determine_action(self._action)
+
+        error = str(exception)
+        level = logging.ERROR
+
+        if isinstance(exception, vol.Invalid):
+            error_desc = "Invalid data"
+
+        elif isinstance(exception, exceptions.TemplateError):
+            error_desc = "Error rendering template"
+
+        elif isinstance(exception, exceptions.Unauthorized):
+            error_desc = "Unauthorized"
+
+        elif isinstance(exception, exceptions.ServiceNotFound):
+            error_desc = "Service not found"
+
+        else:
+            error_desc = "Unexpected error"
+            level = _LOG_EXCEPTION
+
+        self._log(
+            "Error executing script. %s for %s at pos %s: %s",
+            error_desc,
+            action_type,
+            self._step + 1,
+            error,
+            level=level,
+        )
+
+    @abstractmethod
+    async def _async_delay_step(self):
+        """Handle delay."""
+
+    def _start_delay_step(self):
```

Comment:
This name is deceiving as we're not starting anything, we're returning the delay.

Reply:
Maybe prepare ?

Reply:
Good point. I was thinking more "start of the delay/wait_template method" as opposed to "start of the delay/wait_template itself." That change will make it clearer. Will do.